From 3c73d9af5f199e0ff5dabcda7d2894a8a07730d3 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 27 Aug 2015 00:43:02 -0400
Subject: [PATCH] Optimize hot spot with isValidLocation

While this method is simple, its so hot its showing up in samplings.

Replace with an inline check of only the Y value will speed things up,
with no risk when worlds are expected to never be at the 3 billion distance.

Also static the existing isValidLocation to assist with inlining.
---
 src/main/java/net/minecraft/server/World.java | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index c9cecfc..bf9b3ee 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -281,9 +281,11 @@ public abstract class World implements IBlockAccess {
         return this.getType(blockposition1).getBlock();
     }
 
-    private boolean isValidLocation(BlockPosition blockposition) {
+    // EMC start
+    private static boolean isValidLocation(final BlockPosition blockposition) {
         return blockposition.getX() >= -30000000 && blockposition.getZ() >= -30000000 && blockposition.getX() < 30000000 && blockposition.getZ() < 30000000 && blockposition.getY() >= 0 && blockposition.getY() < 256;
     }
+    // EMC end
 
     public boolean isEmpty(BlockPosition blockposition) {
         return this.getType(blockposition).getBlock().getMaterial() == Material.AIR;
@@ -376,7 +378,7 @@ public abstract class World implements IBlockAccess {
             return true;
         }
         // CraftBukkit end
-        if (!this.isValidLocation(blockposition)) {
+        if (blockposition.getY() < 0 || blockposition.getY() >= 256) { // EMC - use a simpler check
             return false;
         } else if (!this.isClientSide && this.worldData.getType() == WorldType.DEBUG_ALL_BLOCK_STATES) {
             return false;
@@ -778,7 +780,7 @@ public abstract class World implements IBlockAccess {
             }
         }
         // CraftBukkit end
-        if (!this.isValidLocation(blockposition)) {
+        if (blockposition.getY() < 0 || blockposition.getY() >= 256) { // EMC
             return Blocks.AIR.getBlockData();
         } else {
             Chunk chunk = this.getChunkAtWorldCoords(blockposition);
-- 
2.7.0

